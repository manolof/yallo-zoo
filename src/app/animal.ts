export interface Animal {
	id: string;
	name: string;
	type: string;
	avatar: string;
}

export interface AnimalsDto {
	items: Animal[];
}
