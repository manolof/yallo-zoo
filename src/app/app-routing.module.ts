import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';
import { AnimalDetailComponent } from './animal-detail/animal-detail.component';
import { AnimalListComponent } from './animal-list/animal-list.component';

const routes: Routes = [
	{ path: 'animals', component: AnimalListComponent },
	{ path: 'animals/:id', component: AnimalDetailComponent },
	{ path: '', redirectTo: '/animals', pathMatch: 'full' },
	{ path: '**', redirectTo: '/animals' },
];

@NgModule({
	imports: [RouterModule.forRoot(routes)],
	exports: [RouterModule],
})
export class AppRoutingModule {}
