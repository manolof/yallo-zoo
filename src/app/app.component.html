<div class="flex flex-col h-screen">
	<app-header></app-header>
	<main class="container mx-auto mb-auto">
		<router-outlet></router-outlet>
	</main>
	<app-footer></app-footer>
</div>
