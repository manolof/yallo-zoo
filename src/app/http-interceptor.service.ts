import {
	HttpInterceptor,
	HttpRequest,
	HttpHandler,
	HttpEvent,
} from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable, of, map } from 'rxjs';

const CACHE = new Map<string, any>();

@Injectable({
	providedIn: 'root',
})
export class HttpInterceptorService implements HttpInterceptor {
	public intercept(
		req: HttpRequest<any>,
		next: HttpHandler,
	): Observable<HttpEvent<any>> {
		if (req.method !== 'GET') {
			return next.handle(req);
		}

		const cachedResponse = CACHE.get(req.urlWithParams);
		if (cachedResponse) {
			return of(cachedResponse);
		}

		return next.handle(req).pipe(
			map(event => {
				CACHE.set(req.urlWithParams, event);
				return event;
			}),
		);
	}
}
