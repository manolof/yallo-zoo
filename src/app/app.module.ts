import { NgModule } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';
import { HttpClientModule, HTTP_INTERCEPTORS } from '@angular/common/http';

import { AppRoutingModule } from './app-routing.module';
import { AppComponent } from './app.component';
import { AnimalCardComponent } from './animal-card/animal-card.component';
import { AnimalListComponent } from './animal-list/animal-list.component';
import { HeaderComponent } from './header/header.component';
import { FooterComponent } from './footer/footer.component';
import { AnimalDetailComponent } from './animal-detail/animal-detail.component';
import { ENV, getEnv } from '../environments/environment.provider';
import { HttpInterceptorService } from './http-interceptor.service';

@NgModule({
	imports: [BrowserModule, HttpClientModule, AppRoutingModule],
	declarations: [
		AppComponent,
		AnimalCardComponent,
		AnimalListComponent,
		AnimalDetailComponent,
		HeaderComponent,
		FooterComponent,
	],
	providers: [
		{ provide: ENV, useFactory: getEnv },
		{
			provide: HTTP_INTERCEPTORS,
			useClass: HttpInterceptorService,
			multi: true,
		},
	],
	bootstrap: [AppComponent],
})
export class AppModule {}
