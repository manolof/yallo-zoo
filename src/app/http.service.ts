import { inject, Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { catchError, map, Observable, of } from 'rxjs';

import { Animal, AnimalsDto } from './animal';
import { ENV } from '../environments/environment.provider';

@Injectable({
	providedIn: 'root',
})
export class HttpService {
	private randomAnimals: Animal[] = [];
	private env = inject(ENV);
	private http = inject(HttpClient);

	public getAnimals(forceRefresh = false): Observable<Animal[]> {
		let params = new HttpParams();
		if (forceRefresh) {
			this.randomAnimals.length = 0;
			params = params.append('nocache', new Date().getTime().toString());
		}

		return this.http.get<AnimalsDto>(this.env.apiUrl, { params }).pipe(
			map(animals => {
				if (this.randomAnimals.length === 0) {
					// Get 6 random animals, for exercise purposes only..
					const shuffled = animals.items.sort(() => 0.5 - Math.random());
					this.randomAnimals = shuffled.slice(0, 6);
				}

				return this.randomAnimals;
			}),
			catchError(this.handleError<Animal[]>('getAnimals', [])),
		);
	}

	public getAnimalDetail(id: string): Observable<Animal> {
		return this.http
			.get<Animal>(`${this.env.apiUrl}/${id}`)
			.pipe(catchError(this.handleError<Animal>('getAnimalDetail')));
	}

	private handleError<T>(operation = 'operation', result?: T) {
		return (error: any): Observable<T> => {
			// TODO: send the error to remote logging infrastructure
			console.error(error); // log to console instead

			// TODO: better job of transforming error for user consumption
			console.error(`${operation} failed: ${error.message}`);

			// Let the app keep running by returning an empty result.
			return of(result as T);
		};
	}
}
