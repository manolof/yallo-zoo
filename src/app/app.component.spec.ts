import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { RouterTestingModule } from '@angular/router/testing';

import { AppComponent } from './app.component';

describe('AppComponent', () => {
	let component: AppComponent;
	let fixture: ComponentFixture<AppComponent>;

	beforeEach(async () => {
		await TestBed.configureTestingModule({
			declarations: [AppComponent],
			imports: [RouterTestingModule],
			schemas: [CUSTOM_ELEMENTS_SCHEMA],
		}).compileComponents();

		fixture = TestBed.createComponent(AppComponent);
		component = fixture.componentInstance;
	});

	it('should create the app', () => {
		expect(component).toBeTruthy();
	});

	it('should match the snapshot', () => {
		fixture.detectChanges();
		expect(fixture).toMatchSnapshot();
	});
});
