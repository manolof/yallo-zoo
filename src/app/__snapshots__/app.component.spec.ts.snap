// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppComponent should match the snapshot 1`] = `
<app-root
  __ngContext__={[Function Number]}
>
  <div
    class="flex flex-col h-screen"
  >
    <app-header />
    <main
      class="container mx-auto mb-auto"
    >
      <router-outlet />
    </main>
    <app-footer />
  </div>
</app-root>
`;
